File: paint_house_dp.py
# Time Complexity : O(n) where n is number of houses
# Space Complexity : O(1) where n is range of the amount
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        n = len(costs)
        costR, costG, costB = costs[n-1][0], costs[n-1][1], costs[n-1][2]
        
        for house in range(n-2, -1, -1):
            tempR = costR
            tempG = costG

            costR = min(costs[house][0] + costG, costs[house][0] + costB)
            costG = min(costs[house][1] + tempR, costs[house][1] + costB)
            costB = min(costs[house][2] + tempR, costs[house][2] + tempG)
            
        return min(costR, costG, costB)

File: paint-house-dp.cpp
// Time Complexity : O(amount * coins.length)
// Space Complexity : O(amount)
// Did this code successfully run on Leetcode : Yes

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    static int minCost(vector<vector<int>>& costs) {
        if (costs.empty() || costs[0].empty())
            return 0;

        for (int i = 1; i < costs.size(); i++) {
            costs[i][0] += min(costs[i - 1][1], costs[i - 1][2]);
            costs[i][1] += min(costs[i - 1][0], costs[i - 1][2]);
            costs[i][2] += min(costs[i - 1][0], costs[i - 1][1]);
        }

        int m = costs.size() - 1;
        return min(min(costs[m][0], costs[m][1]), costs[m][2]);
    }
};